src/libtls/tls_server.c

```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		key = lib->credmgr->get_private(lib->credmgr, *v, this->server,
+										this->server_auth);
+		if (key)
+		{
+			type = *v;
+			break;
+		}
+		key_types->remove_at(key_types, enumerator);
+		free(v);
+	}
+	enumerator->destroy(enumerator);
 	if (!key)
 	{
 		DBG1(DBG_TLS, "no usable TLS server certificate found for '%Y'",
 			 this->server);
+		key_types->destroy(key_types);
 		return FALSE;
 	}
 
-	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	if (version_max >= TLS_1_3)
 	{
-		/* currently no support to derive key based on client supported
-		 * signature schemes */
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 												  		count, KEY_ANY);
-		if (!this->suite)
-		{
-			DBG1(DBG_TLS, "received cipher suites unacceptable");
-			return FALSE;
-		}
 	}
 	else
 	{
-		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
-												  		count,
-												  		key->get_type(key));
-		if (!this->suite)
-		{	/* no match for this key, try to find another type */
-			if (key->get_type(key) == KEY_ECDSA)
-			{
-				type = KEY_RSA;
-			}
-			else
-			{
-				type = KEY_ECDSA;
-			}
-			key->destroy(key);
-
+		if (type)
+		{
 			this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 															count, type);
-			if (!this->suite)
-			{
-				DBG1(DBG_TLS, "received cipher suites unacceptable");
-				return FALSE;
-			}
-			this->server_auth->destroy(this->server_auth);
-			this->server_auth = auth_cfg_create();
-			key = lib->credmgr->get_private(lib->credmgr, type, this->server,
-											this->server_auth);
-			if (!key)
-			{
-				DBG1(DBG_TLS, "received cipher suites unacceptable");
-				return FALSE;
+		}
+		if (!this->suite)
+		{
+			enumerator = key_types->create_enumerator(key_types);
+			while (enumerator->enumerate(enumerator, &k, &v))
+			{	/* find a cipher suite that supports the key */
+				type = *v;
+				key_types->remove_at(key_types, enumerator);
+				free(v);
+
+				if (!key)
+				{
+					this->server_auth->destroy(this->server_auth);
+					this->server_auth = auth_cfg_create();
+					key = lib->credmgr->get_private(lib->credmgr, type,
+									 				this->server,
+													this->server_auth);
+					if (!key)
+					{
+						continue;
+					}
+				}
+				this->suite = this->crypto->select_cipher_suite(this->crypto,
+																suites, count,
+																type);
+				if (this->suite)
+				{
+					break;
+				}
+				key->destroy(key);
+				key = NULL;
 			}
+			enumerator->destroy(enumerator);
 		}
 	}
+	if (!this->suite || !key)
+	{
+		DBG1(DBG_TLS, "received cipher suites unacceptable");
+		key_types->destroy(key_types);
+		return FALSE;
+	}
+	enumerator = key_types->create_enumerator(key_types);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		key_types->remove_at(key_types, enumerator);
+		free(v);
+	}
+	enumerator->destroy(enumerator);
+	key_types->destroy(key_types);
```

Comment:
You could use `destroy_function()` and pass `free()` as callback. But using an array will be simpler.